File: kotlin/src/main/kotlin/Benchmark.kt
import kotlin.system.exitProcess

abstract class Benchmark {
    abstract fun run(iterationId: Int)

    abstract fun checksum(): UInt

    open fun prepare() {}

    abstract fun name(): String

    open fun warmupIterations(): Long =
        Helper.optConfigI64(name(), "warmup_iterations")
            ?: maxOf((iterations() * 0.2).toLong(), 1L)

    open fun warmup() {
        repeat(warmupIterations().toInt()) { run(it) }
    }

    open fun runAll() {
        repeat(iterations().toInt()) { run(it) }
    }

    open fun configVal(fieldName: String): Long = Helper.configI64(name(), fieldName)

    fun configInt(fieldName: String): Int = configVal(fieldName).toInt()

    fun configStr(fieldName: String): String = Helper.configS(name(), fieldName)

    open fun iterations(): Long = configVal("iterations")

    open fun expectedChecksum(): Long = configVal("checksum")

    companion object {
        private val benchmarkMap = mutableMapOf<String, () -> Benchmark>()

        fun registerBenchmark(
            name: String,
            factory: () -> Benchmark,
        ) {
            if (benchmarkMap.containsKey(name)) {
                System.err.println("Warning: Benchmark with name '$name' already registered. Skipping.")
                return
            }
            benchmarkMap[name] = factory
        }

        fun all(singleBench: String? = null) {
            var summaryTime = 0.0
            var ok = 0
            var fails = 0

            for (benchName in Helper.order) {
                if (singleBench != null && !benchName.contains(singleBench, ignoreCase = true)) {
                    continue
                }

                val factory = benchmarkMap[benchName]
                if (factory == null) {
                    println("Warning: Benchmark '$benchName' defined in config but not found in code")
                    continue
                }

                val bench = factory()

                Helper.reset()

                bench.prepare()
                bench.warmup()
                System.gc()

                Helper.reset()

                val startTime = System.nanoTime()
                bench.runAll()
                val timeDelta2 = (System.nanoTime() - startTime) / 1_000_000_000.0

                System.gc()
                Thread.sleep(1)
                System.gc()

                print("$benchName: ")
                if (bench.checksum().toLong() == bench.expectedChecksum()) {
                    print("OK ")
                    ok++
                } else {
                    print("ERR[actual=${bench.checksum()}, expected=${bench.expectedChecksum()}] ")
                    fails++
                }

                println("in %.3fs".format(timeDelta2))
                summaryTime += timeDelta2
            }

            println("Summary: %.4fs, %d, %d, %d".format(summaryTime, ok + fails, ok, fails))

            if (fails > 0) {
                exitProcess(1)
            }
        }
    }
}

File: fsharp/Benchmark.fs
namespace Benchmarks

open System
open System.Collections.Generic
open System.Diagnostics
open System.IO
open System.Reflection
open System.Text.Json
open System.Threading

[<AbstractClass>]
type Benchmark() =
    abstract member Run: int64 -> unit
    abstract member Checksum: uint32
    abstract member Prepare: unit -> unit
    default _.Prepare() = ()

    abstract member Warmup: unit -> unit
    default this.Warmup() =
        let prepareIters = this.WarmupIterations
        for i in 0L .. prepareIters - 1L do
            this.Run(i)

    member val TimeDelta = 0.0 with get, set

    member this.Name = this.GetType().Name

    member this.WarmupIterations : int64 =
        let className = this.Name
        let mutable benchObj = JsonElement()
        if Helper.Config.TryGetProperty(className, &benchObj) then
            let mutable warmupProp = JsonElement()
            if benchObj.TryGetProperty("warmup_iterations", &warmupProp) then
                warmupProp.GetInt64()
            else
                let iters = this.Iterations
                max (int64 (double iters * 0.2)) 1L
        else
            let iters = this.Iterations
            max (int64 (double iters * 0.2)) 1L

    member this.RunAll() =
        let iters = this.Iterations
        for i in 0L .. iters - 1L do
            this.Run(i)

    member this.ConfigVal(fieldName: string) : int64 =
        Helper.Config_i64(this.Name, fieldName)

    member this.Iterations : int64 = this.ConfigVal("iterations")

    member this.ExpectedChecksum : int64 = this.ConfigVal("checksum")

    static member All(?singleBench: string) =
        let singleBench = defaultArg singleBench null

        let now = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds()
        printfn "start: %d" now

        let results = Dictionary<string, double>()
        let mutable summaryTime = 0.0
        let mutable ok = 0
        let mutable fails = 0

        let assembly = Assembly.GetExecutingAssembly()
        let benchmarkTypes = 
            assembly.GetTypes()
            |> Array.filter (fun t -> 
                t.IsSubclassOf(typeof<Benchmark>) && not t.IsAbstract)

        let benchmarks = 
            benchmarkTypes
            |> Array.map (fun t -> Activator.CreateInstance(t) :?> Benchmark)
            |> Array.toList

        for benchmark in benchmarks do
            let className = benchmark.Name

            if not (String.IsNullOrEmpty singleBench) && 
               not (className.ToLower().Contains(singleBench.ToLower())) then
                ()
            else
                let mutable benchConfig = JsonElement()
                if not (Helper.Config.TryGetProperty(className, &benchConfig)) then
                    printfn "Skipping %s - no config in test.js" className
                else
                    printf "%s: " className

                    Helper.Reset()

                    try
                        benchmark.Prepare()
                        benchmark.Warmup()

                        Helper.Reset()

                        let stopwatch = Stopwatch.StartNew()
                        benchmark.RunAll()
                        stopwatch.Stop()

                        let timeDelta = stopwatch.Elapsed.TotalSeconds
                        benchmark.TimeDelta <- timeDelta
                        results.[className] <- timeDelta

                        GC.Collect()
                        Thread.Sleep(0)
                        GC.Collect()

                        let actual = benchmark.Checksum
                        let expected = uint32 benchmark.ExpectedChecksum

                        if actual = expected then
                            printf "OK "
                            ok <- ok + 1
                        else
                            printf "ERR[actual=%d, expected=%d] " actual expected
                            fails <- fails + 1

                        printfn "in %.3fs" timeDelta
                        summaryTime <- summaryTime + timeDelta
                    with
                    | ex -> 
                        printfn "ERROR: %s" ex.Message
                        fails <- fails + 1

        try
            let json = "{" + String.Join(",", results |> Seq.map (fun kv -> sprintf "\"%s\":%f" kv.Key kv.Value)) + "}"
            File.WriteAllText("/tmp/results.js", json)
        with
        | ex -> printfn "Error saving results: %s" ex.Message

        printfn "Summary: %.4fs, %d, %d, %d" summaryTime (ok + fails) ok fails

        File.WriteAllText("/tmp/recompile_marker", "RECOMPILE_MARKER_0")

        if fails > 0 || ok = 0 then
            Environment.Exit(1)